zerver/migrations/0371_create_role_based_system_groups.py

```diff
@@ -0,0 +1,121 @@
+            "description": "Everyone in this organization, including guests",
+        },
+    }
+
+    for realm in realms:
+        role_system_groups_dict = {}
+        for role in SYSTEM_USER_GROUP_ROLE_MAP.keys():
+            user_group_params = SYSTEM_USER_GROUP_ROLE_MAP[role]
+            user_group = UserGroup(
+                name=user_group_params["name"],
+                description=user_group_params["description"],
+                realm=realm,
+                is_system_group=True,
+            )
+            role_system_groups_dict[role] = user_group
+
+        full_members_system_group = UserGroup(
+            name="@role:fullmembers",
+            description="Members of this organization, not including new accounts and guests",
+            realm=realm,
+            is_system_group=True,
+        )
+        everyone_on_internet_system_group = UserGroup(
+            name="@role:internet",
+            description="Everyone on the Internet",
+            realm=realm,
+            is_system_group=True,
+        )
+
+        system_user_groups_list = [
+            role_system_groups_dict[UserProfile.ROLE_REALM_OWNER],
+            role_system_groups_dict[UserProfile.ROLE_REALM_ADMINISTRATOR],
+            role_system_groups_dict[UserProfile.ROLE_MODERATOR],
+            full_members_system_group,
+            role_system_groups_dict[UserProfile.ROLE_MEMBER],
+            role_system_groups_dict[UserProfile.ROLE_GUEST],
+            everyone_on_internet_system_group,
+        ]
+
+        UserGroup.objects.bulk_create(system_user_groups_list)
+
+        subgroup_objects = []
+        for idx, supergroup in enumerate(system_user_groups_list[1:]):
+            subgroup_objects.append(
+                GroupGroupMembership(subgroup=system_user_groups_list[idx], supergroup=supergroup)
+            )
+
+        GroupGroupMembership.objects.bulk_create(subgroup_objects)
+
+        users = UserProfile.objects.filter(realm=realm)
+        group_membership_obejcts = []
+        for user in users:
+            system_group = role_system_groups_dict[user.role]
+            group_membership_obejcts.append(
+                UserGroupMembership(user_profile=user, user_group=system_group)
+            )
+
+            if (
+                user.role == UserProfile.ROLE_MEMBER
+                and (timezone_now() - user.date_joined).days >= realm.waiting_period_threshold
+            ):
+                group_membership_obejcts.append(
+                    UserGroupMembership(user_profile=user, user_group=full_members_system_group)
+                )
+
+        UserGroupMembership.objects.bulk_create(group_membership_obejcts)
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ("zerver", "0370_realm_enable_spectator_access"),
+    ]
+
+    operations = [
+        migrations.RunPython(
+            create_role_based_system_groups, reverse_code=migrations.RunPython.noop, elidable=True
```

Comment:
This migration seems like it might be somewhat slow, since it's O(totalusers).  @mateuszmandera I'm curious for your thoughts, but perhaps I think we can address that with:
* Setting `atomic=False`
* Doing each realm inside a transaction; that should prevent us from ending up with a realm partially created if we have a failure or ctrl+C or whatever.
* Planning to run this migration only after we've started creating these for new realms.  I think given that plan, we should make it check that the relevant user groups don't already exist in the target realm, and skip the realm if so.

Reply:
@timabbott @sahil839 Yeah, this plan sounds rights to me - it's typical for us to batch by realm in these kinds of migrations, with each realm being processed in its own atomic transaction

Reply:
Sorry, but it is not clear to me what will be the advantage of the third point of running the migration after we have started creating groups for new realms. Otherwise, sounds good to me.

Reply:
OK, added `atomic=False`, and this block at the top of the migration:
```
+        with transaction.atomic():
+            if UserGroup.objects.filter(
+                realm=realm, name="@role:internet", is_system_group=True
+            ).exists():
+                # Handle the case where we are rerunning after a
+                # failure, and had already processed this realm.
+                continue
```

Reply:
I also synced the loop to make `GroupGroupMembership` objects to use the `for supergroup in remaining_groups` loop we use in `actions.py`, and added a comment.